File: biomed-maintenance-app/fix_numbers.js
import { createClient } from '@supabase/supabase-js';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const envPath = path.resolve(__dirname, '.env');
if (fs.existsSync(envPath)) {
    const envConfig = fs.readFileSync(envPath, 'utf-8');
    envConfig.split('\n').forEach(line => {
        const match = line.match(/^\s*([\w.-]+)\s*=\s*(.*)?\s*$/);
        if (match) process.env[match[1]] = match[2].replace(/^['"]|['"]$/g, '');
    });
}

const supabaseUrl = process.env.SUPABASE_URL || 'https://gzdspkhpxkibjxbfdeuc.supabase.co';
const supabaseKey = process.env.SUPABASE_SERVICE_ROLE_KEY;
if (!supabaseKey) throw new Error('SUPABASE_SERVICE_ROLE_KEY is missing. Set it in your .env file: SUPABASE_SERVICE_ROLE_KEY=your-key');
const supabase = createClient(supabaseUrl, supabaseKey);

async function fix() {
    console.log("🔍 Buscando fantasmas con .0...");
    const { data, error } = await supabase
        .from('equipments')
        .select('id, id_unico, numero_serie');
    
    if (error) throw error;

    const toDelete = data.filter(e => 
        String(e.id_unico || '').endsWith('.0') || 
        String(e.numero_serie || '').endsWith('.0')
    ).map(e => e.id);

    if (toDelete.length > 0) {
        console.log(`🗑️ Purgando ${toDelete.length} registros con formato .0...`);
        for (let i = 0; i < toDelete.length; i += 50) {
            await supabase.from('equipments').delete().in('id', toDelete.slice(i, i + 50));
        }
        console.log("✅ Purga completa.");
    }

    // RE-SYNC with Deduplication
    console.log("📡 Resincronizando datos limpios...");
    const csvContent = fs.readFileSync('master_inventory_bombs_final_431.csv', 'utf8');
    const lines = csvContent.split('\n').filter(l => l.trim());
    
    const uniqueMap = new Map();
    lines.slice(1).forEach(l => {
        const p = l.split(',');
        const s = String(p[0] || '').trim();
        const idu = String(p[1] || '').trim() || s;
        if (!idu) return;

        uniqueMap.set(idu, {
            numero_serie: s,
            id_unico: idu,
            equipo: p[2] || 'BOMBA DE INFUSIÓN',
            marca: p[3] || 'ICU MEDICAL',
            modelo: p[4] || 'N/A',
            servicio: p[5] || 'N/A',
            ubicacion: p[7] || 'N/A',
            estado: 'BUENO',
            riesgo: 'IIb',
            propietario: 'HUSJ (COMODATO)'
        });
    });

    const records = Array.from(uniqueMap.values());
    console.log(`📊 Registros únicos para subir: ${records.length}`);

    for (let i = 0; i < records.length; i += 50) {
        const chunk = records.slice(i, i + 50);
        const { error: upError } = await supabase.from('equipments').upsert(chunk, { onConflict: 'id_unico' });
        if (upError) {
            console.error(`❌ Error upsert en bloque ${i}:`, upError.message);
        } else {
            process.stdout.write('.');
        }
    }
    console.log("\n✨ Sincronización final COMPLETADA sin duplicados ni .0.");
}

fix().catch(console.error);

File: _dev_archive/fix_numbers.py
import urllib.request
import json
import ssl
import pandas as pd

url_base = "https://gzdspkhpxkibjxbfdeuc.supabase.co/rest/v1"
apikey = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imd6ZHNwa2hweGtpYmp4YmZkZXVjIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc3NDg4MjM1OSwiZXhwIjoyMDkwNDU4MzU5fQ.yCjghk-O_HOttUJBDhJ_LvkcdbLN97GMyrSwyV574Tw"
headers = {
    "apikey": apikey,
    "Authorization": f"Bearer {apikey}",
    "Content-Type": "application/json"
}
context = ssl._create_unverified_context()

def api_call(path, method="GET", body=None, extra_headers=None):
    full_url = f"{url_base}{path}"
    h = headers.copy()
    if extra_headers: h.update(extra_headers)
    data = json.dumps(body).encode("utf-8") if body else None
    req = urllib.request.Request(full_url, data=data, headers=h, method=method)
    with urllib.request.urlopen(req, context=context) as response:
        res_data = response.read()
        return json.loads(res_data) if res_data else None

# 1. FETCH ALL
print("Buscando registros con .0 para purgar...")
current = api_call("/equipments?select=id,id_unico,numero_serie")
to_delete = [item["id"] for item in current if str(item["id_unico"]).endswith(".0") or str(item["numero_serie"]).endswith(".0")]

if to_delete:
    print(f"Purgando {len(to_delete)} registros con formato incorrecto (.0)...")
    for i in range(0, len(to_delete), 50):
        ids_str = ",".join([f'"{id}"' for id in to_delete[i:i+50]])
        api_call(f"/equipments?id=in.({ids_str})", method="DELETE")
    print("Purga completada.")
else:
    print("No se encontraron registros con .0")

# 2. RE-UPSERT CLEAN DATA
df = pd.read_csv("master_inventory_bombs_final_431.csv")
print("Sincronizando 431 equipos limpios...")
payload = []
for _, row in df.iterrows():
    s = str(row["Serial"])
    payload.append({
        "id_unico": str(row["Id_Unico"]),
        "numero_serie": s,
        "equipo": str(row["Equipo"]),
        "marca": str(row["Marca"]),
        "modelo": str(row["Modelo"]),
        "servicio": str(row["Servicio"]),
        "ubicacion": str(row["Nueva_Ubicacion"]),
        "estado": "BUENO",
        "riesgo": "IIb",
        "propietario": "HUSJ (COMODATO)"
    })

for i in range(0, len(payload), 50):
    batch = payload[i:i+50]
    api_call("/equipments?on_conflict=id_unico", method="POST", body=batch, extra_headers={"Prefer": "resolution=merge-duplicates"})
    print(".", end="", flush=True)

print("\n✨ ¡Sincronización FINAL sin .0 completada!")
